File: Code/2/6/2610/2610.cpp
/******************************************************************************
 * 文件名: 2610
 * 作者: BOLUN XU
 * 创建日期: 2025
 * 版本: 1.0
 * 描述: 转换二维数组。
 * 时间复杂度： O(N^2)
 * 空间复杂度： O(N)
 ******************************************************************************/
#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> findMatrix(vector<int>& nums) {
        vector<vector<int>> ans;
        unordered_map<int, int> cnt;

        // 统计每个元素的出现次数
        for (int num : nums) {
            cnt[num]++;
        }

        while (!cnt.empty()) {
            vector<int> row;
            for (auto& pair : cnt) {
                row.push_back(pair.first);
            }
            ans.push_back(row);

            // cnt 中的每个元素的出现次数都减一
            for (int x : row) {
                cnt[x]--;
                if (cnt[x] == 0) {
                    cnt.erase(x);
                }
            }
        }

        return ans;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {1, 3, 4, 1, 2, 3, 1};
    vector<vector<int>> result = sol.findMatrix(nums);

    // 输出结果
    for (const auto& row : result) {
        for (int num : row) {
            cout << num << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Code/2/6/2610/2610.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
文件名: PersonalLeetCode 2610.py
作者: Bolun Xu
创建日期: 2025/3/19
版本: 1.0
描述: 转换二维数组。
时间复杂度： O(N)
空间复杂度： O(N)
"""
from collections import Counter
from typing import List


class Solution:
    def findMatrix(self, nums: List[int]) -> List[List[int]]:
        ans = []
        cnt = Counter(nums)  # 统计每个元素的出现次数
        # print(f"DEBUG: cnt = {cnt}")
        while cnt:  # 还有剩余元素
            row = list(cnt)  # list(cnt)会将cnt中的 键 转换为一个列表
            # print(f"DEBUG: row = {row}")
            ans.append(row)
            # print(f"DEBUG: ans = {ans}")
            # cnt 中的每个元素的出现次数都减一
            for x in row:
                cnt[x] -= 1
                if cnt[x] == 0:
                    del cnt[x]  # 去掉出现次数为 0 的元素
        return ans


if __name__ == '__main__':
    sol = Solution()
    print(sol.findMatrix([1, 3, 4, 1, 2, 3, 1]))
